File: sdk/rust/src/loader.rs
//! Test suite loaders for various formats.
//!
//! The YAML loader reads a `metadata.yaml` that lists test-case entries.
//! Each entry names a plan binary and optional input/expected CSV files.
//! Expected CSV files use typed headers in the form `colname:type` and are
//! located in an `expected/` subdirectory next to the metadata file.

use crate::error::{ComplianceError, Result};
use crate::table_data::{Column, DataType, TableData};
use crate::test_suite::{SimpleTestSuite, TestCase, TestSuite, TestSuiteMetadata};
use serde::Deserialize;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

/// Trait for loading test suites.
pub trait TestSuiteLoader {
    fn load(&self, path: &Path) -> Result<Box<dyn TestSuite>>;
    fn supports(&self, path: &Path) -> bool;
}

/// Loads test suites from YAML metadata files.
pub struct YamlTestSuiteLoader;

impl YamlTestSuiteLoader {
    pub fn new() -> Self { Self }
}

impl Default for YamlTestSuiteLoader {
    fn default() -> Self { Self::new() }
}

// ── YAML schema ─────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
struct TestSuiteDefinition {
    name: String,
    version: String,
    #[serde(default)]
    description: String,
    #[serde(rename = "testCases", default)]
    test_cases: Vec<TestCaseDefinition>,
}

#[derive(Debug, Deserialize)]
struct TestCaseDefinition {
    id: String,
    #[serde(default)]
    description: String,
    /// Relative path to the binary Substrait plan file.
    #[serde(rename = "planBinary")]
    plan_binary: String,
    /// Optional relative path to the expected-output CSV.
    /// Defaults to `expected/<id>.csv` when omitted.
    #[serde(rename = "expectedOutput")]
    expected_output: Option<String>,
}

// ── Loader impl ──────────────────────────────────────────────────────────────

impl TestSuiteLoader for YamlTestSuiteLoader {
    fn load(&self, path: &Path) -> Result<Box<dyn TestSuite>> {
        let content = fs::read_to_string(path)?;
        let def: TestSuiteDefinition = serde_yaml::from_str(&content)?;

        let base_dir = path.parent().ok_or_else(||
            ComplianceError::Loading("Invalid path".to_string())
        )?;

        let mut test_cases = Vec::new();
        for tc_def in def.test_cases {
            let plan_path = base_dir.join(&tc_def.plan_binary);
            let plan_bytes = fs::read(&plan_path).map_err(|e| {
                ComplianceError::Loading(format!("plan {}: {}", plan_path.display(), e))
            })?;

            let mut tc = TestCase::new(tc_def.id.clone(), tc_def.description, plan_bytes);

            // Resolve expected-output path: explicit field or default location.
            let csv_path = match &tc_def.expected_output {
                Some(rel) => base_dir.join(rel),
                None      => base_dir.join("expected").join(format!("{}.csv", tc_def.id)),
            };
            if csv_path.exists() {
                tc.expected_output = Some(load_csv(&csv_path)?);
            }

            test_cases.push(tc);
        }

        let metadata = TestSuiteMetadata {
            name: def.name.clone(),
            version: def.version,
            description: def.description,
        };

        Ok(Box::new(SimpleTestSuite::new(def.name, test_cases, metadata)))
    }

    fn supports(&self, path: &Path) -> bool {
        path.extension()
            .and_then(|e| e.to_str())
            .map(|e| e == "yaml" || e == "yml")
            .unwrap_or(false)
    }
}

// ── CSV / expected-output parsing ────────────────────────────────────────────

/// Parse a pipe-delimited CSV with an optional typed header row.
///
/// Typed header format: `colname:type|colname:type|...`
/// Untyped header: absent — columns are named `column_1`, `column_2`, …
pub fn load_csv(path: &Path) -> Result<TableData> {
    let text = fs::read_to_string(path).map_err(|e| {
        ComplianceError::Loading(format!("expected CSV {}: {}", path.display(), e))
    })?;

    let lines: Vec<&str> = text.lines().collect();
    if lines.is_empty() {
        return Ok(TableData::new(vec![], vec![]));
    }

    // Detect typed header: every field must contain ':'
    let first = lines[0];
    let fields: Vec<&str> = first.split('|').collect();
    let has_typed_header = fields.iter().all(|f| f.contains(':'));

    let (columns, data_start) = if has_typed_header {
        let cols = fields.iter().map(|f| {
            let (name, type_str) = f.split_once(':').unwrap_or((f, "varchar"));
            Column::new(name.trim(), parse_data_type(type_str.trim()))
        }).collect();
        (cols, 1)
    } else {
        // No header: default column names, all varchar
        let ncols = fields.len();
        let cols = (1..=ncols)
            .map(|i| Column::new(format!("column_{}", i), DataType::Varchar))
            .collect();
        (cols, 0)
    };

    let mut rows: Vec<Vec<String>> = Vec::new();
    for line in &lines[data_start..] {
        let line = line.trim();
        if line.is_empty() { continue; }
        let row: Vec<String> = line.split('|').map(|v| v.trim().to_string()).collect();
        rows.push(row);
    }

    Ok(TableData::new(columns, rows))
}

fn parse_data_type(s: &str) -> DataType {
    match s.to_lowercase().as_str() {
        "integer" | "int" | "int32" | "i32" | "smallint" | "int4" => DataType::Integer,
        "bigint" | "int64" | "i64"                                  => DataType::Bigint,
        "double" | "fp64" | "float8" | "numeric" | "decimal"        => DataType::Double,
        "boolean" | "bool"                                           => DataType::Boolean,
        "date" | "timestamp"                                         => DataType::Date,
        _                                                            => DataType::Varchar,
    }
}

// ── Input-data CSV ────────────────────────────────────────────────────────────

/// Load a directory of named input CSVs into a `HashMap<name, TableData>`.
/// Each file `tablename.csv` becomes the entry keyed by `tablename`.
pub fn load_input_data(dir: &Path) -> Result<HashMap<String, TableData>> {
    let mut map = HashMap::new();
    if !dir.exists() { return Ok(map); }
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) == Some("csv") {
            let name = path.file_stem()
                .and_then(|s| s.to_str())
                .unwrap_or("unknown")
                .to_string();
            map.insert(name, load_csv(&path)?);
        }
    }
    Ok(map)
}

File: sdk/go/loader.go
package compliance

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

// TestSuiteLoader is the interface for loading test suites
type TestSuiteLoader interface {
	Load(path string) (*TestSuite, error)
	Supports(path string) bool
}

// YAMLTestSuiteLoader loads test suites from YAML files
type YAMLTestSuiteLoader struct{}

// NewYAMLTestSuiteLoader creates a new YAML loader
func NewYAMLTestSuiteLoader() *YAMLTestSuiteLoader {
	return &YAMLTestSuiteLoader{}
}

// ── YAML schema ───────────────────────────────────────────────────────────────

type testSuiteYAML struct {
	Name        string          `yaml:"name"`
	Version     string          `yaml:"version"`
	Description string          `yaml:"description"`
	TestCases   []testCaseYAML  `yaml:"testCases"`
}

type testCaseYAML struct {
	ID             string `yaml:"id"`
	Description    string `yaml:"description"`
	PlanBinary     string `yaml:"planBinary"`
	// Explicit override for expected-output CSV path (relative to YAML dir).
	// Defaults to expected/<id>.csv when omitted.
	ExpectedOutput string `yaml:"expectedOutput"`
}

// ── Loader impl ───────────────────────────────────────────────────────────────

// Load loads a test suite from a YAML metadata file.
// For each test case it reads the plan binary and, if present, the
// expected-output CSV from expected/<id>.csv (or the explicit path in the YAML).
func (l *YAMLTestSuiteLoader) Load(path string) (*TestSuite, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read metadata file: %w", err)
	}

	var def testSuiteYAML
	if err := yaml.Unmarshal(data, &def); err != nil {
		return nil, fmt.Errorf("failed to parse metadata: %w", err)
	}

	baseDir := filepath.Dir(path)

	suite := NewTestSuite(TestSuiteMetadata{
		Name:        def.Name,
		Version:     def.Version,
		Description: def.Description,
	})

	for _, tc := range def.TestCases {
		// Read plan bytes
		planPath := filepath.Join(baseDir, tc.PlanBinary)
		planBytes, err := os.ReadFile(planPath)
		if err != nil {
			return nil, fmt.Errorf("test case %s: failed to read plan %s: %w", tc.ID, planPath, err)
		}

		testCase := NewTestCase(tc.ID, planBytes).WithDescription(tc.Description)

		// Resolve expected-output CSV
		csvPath := filepath.Join(baseDir, "expected", tc.ID+".csv")
		if tc.ExpectedOutput != "" {
			csvPath = filepath.Join(baseDir, tc.ExpectedOutput)
		}
		if _, statErr := os.Stat(csvPath); statErr == nil {
			expected, csvErr := LoadCSV(csvPath)
			if csvErr != nil {
				return nil, fmt.Errorf("test case %s: %w", tc.ID, csvErr)
			}
			testCase.WithExpectedOutput(expected)
		}

		suite.AddTestCase(testCase)
	}

	return suite, nil
}

// Supports checks if the loader supports the given file
func (l *YAMLTestSuiteLoader) Supports(path string) bool {
	ext := filepath.Ext(path)
	return ext == ".yaml" || ext == ".yml"
}

// ── CSV parsing ───────────────────────────────────────────────────────────────

// LoadCSV parses a pipe-delimited CSV file into a TableData.
//
// Typed header format (detected automatically):
//   colname:type|colname:type|...
//
// If every field in the first line contains ':', it is treated as a typed
// header and consumed.  Otherwise columns are named column_1, column_2, …
// and the first line is treated as a data row.
func LoadCSV(path string) (*TableData, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("failed to open CSV %s: %w", path, err)
	}
	defer f.Close()

	var lines []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" {
			lines = append(lines, line)
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("reading CSV %s: %w", path, err)
	}
	if len(lines) == 0 {
		return NewTableData(nil), nil
	}

	// Detect typed header
	fields := strings.Split(lines[0], "|")
	hasTypedHeader := true
	for _, f := range fields {
		if !strings.Contains(f, ":") {
			hasTypedHeader = false
			break
		}
	}

	var columns []ColumnMetadata
	dataStart := 0

	if hasTypedHeader {
		for _, f := range fields {
			f = strings.TrimSpace(f)
			idx := strings.LastIndex(f, ":")
			name := strings.TrimSpace(f[:idx])
			typStr := strings.TrimSpace(f[idx+1:])
			columns = append(columns, ColumnMetadata{Name: name, Type: normalizeCSVType(typStr)})
		}
		dataStart = 1
	} else {
		for i := range fields {
			columns = append(columns, ColumnMetadata{
				Name: fmt.Sprintf("column_%d", i+1),
				Type: "string",
			})
		}
	}

	td := NewTableData(columns)
	for _, line := range lines[dataStart:] {
		parts := strings.Split(line, "|")
		row := make(Row, len(parts))
		for i, p := range parts {
			row[i] = strings.TrimSpace(p)
		}
		td.AddRow(row)
	}
	return td, nil
}

// normalizeCSVType maps CSV type strings to canonical names.
func normalizeCSVType(s string) string {
	switch strings.ToLower(strings.TrimSpace(s)) {
	case "integer", "int", "int32", "i32", "smallint", "int4":
		return "integer"
	case "bigint", "int64", "i64":
		return "bigint"
	case "double", "fp64", "float8", "numeric", "decimal", "number":
		return "double"
	case "boolean", "bool":
		return "boolean"
	default:
		return "string"
	}
}

// ── Convenience functions ─────────────────────────────────────────────────────

// LoadPlan loads a Substrait plan from a file.
func LoadPlan(path string) ([]byte, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read plan file: %w", err)
	}
	return data, nil
}

// LoadTestSuite is a convenience function to load a test suite.
func LoadTestSuite(path string) (*TestSuite, error) {
	return NewYAMLTestSuiteLoader().Load(path)
}

// AutoTestSuiteLoader tries multiple loaders
type AutoTestSuiteLoader struct {
	loaders []TestSuiteLoader
}

// NewAutoTestSuiteLoader creates a new auto-detecting loader
func NewAutoTestSuiteLoader() *AutoTestSuiteLoader {
	return &AutoTestSuiteLoader{
		loaders: []TestSuiteLoader{NewYAMLTestSuiteLoader()},
	}
}

// RegisterLoader adds a custom loader
func (l *AutoTestSuiteLoader) RegisterLoader(loader TestSuiteLoader) {
	l.loaders = append(l.loaders, loader)
}

// Load loads using the first compatible loader
func (l *AutoTestSuiteLoader) Load(path string) (*TestSuite, error) {
	for _, loader := range l.loaders {
		if loader.Supports(path) {
			return loader.Load(path)
		}
	}
	return nil, fmt.Errorf("no loader found for file: %s", path)
}

// Supports checks if any loader supports the file
func (l *AutoTestSuiteLoader) Supports(path string) bool {
	for _, loader := range l.loaders {
		if loader.Supports(path) {
			return true
		}
	}
	return false
}
